src/core/border_router/routing_manager.cpp

```diff
@@ -1410,16 +1436,24 @@ void RoutingManager::InvalidateDiscoveredPrefixes(const Ip6::Prefix *aPrefix, bo
 
     for (const ExternalPrefix &prefix : mDiscoveredPrefixes)
     {
+        bool isAdvertisedLocalOnLinkPrefix =
+            mIsAdvertisingLocalOnLinkPrefix && prefix.mIsOnLinkPrefix && mLocalOnLinkPrefix == prefix.mPrefix;
+
         if (
             // Invalidate specified prefix
             (aPrefix != nullptr && prefix.mPrefix == *aPrefix && prefix.mIsOnLinkPrefix == aIsOnLinkPrefix) ||
             // Invalidate expired prefix
             (prefix.GetExpireTime() <= now) ||
             // Invalidate Local OMR prefixes
             (!prefix.mIsOnLinkPrefix &&
-             (mAdvertisedOmrPrefixes.Contains(prefix.mPrefix) || NetworkDataContainsOmrPrefix(prefix.mPrefix))))
+             (mAdvertisedOmrPrefixes.Contains(prefix.mPrefix) || NetworkDataContainsOmrPrefix(prefix.mPrefix))) ||
+            // Remove local on-link prefix if the BR is advertising on-link prefix
+            isAdvertisedLocalOnLinkPrefix)
         {
-            UnpublishExternalRoute(prefix.mPrefix);
+            if (!isAdvertisedLocalOnLinkPrefix)
+            {
+                UnpublishExternalRoute(prefix.mPrefix);
+            }
```

Comment:
It can be a bit cleaner that we want to remove the cache of discovered local on-link prefix but don't want to remove the route for it:
```
if (... || (!prefix.mIsOnLinkPrefix &&
             (mAdvertisedOmrPrefixes.Contains(prefix.mPrefix) || NetworkDataContainsOmrPrefix(prefix.mPrefix))))) {
    UnpublishExternalRoute(prefix.mPrefix);
} else if (isAdvertisedLocalOnLinkPrefix) {
    // Do nothing to remove local on-link prefix from `mDiscoveredPrefixes` but keep the external route. 
}
else
{
    // ......
}
```

Reply:
Your pseudo-code is actually incorrect, unless the `if (isAdvertisedLocalOnLinkPrefix)` case is moved to be the first condition. 

I think it looks risky to change the code. Inclined to keep the current code as is since it's already clear enough.